### PR TITLE
[Model Averaging] Support hierarchical model averaging

### PR DESCRIPTION
**Pull request by wayi1**
Implement hierarchical model averaging proposed in https://github.com/pytorch/pytorch/issues/71325.

Unit tests are added. Since I don't have access to 4-GPU machines in open-source environment, expect that the branch with the prefix of `ci-all` can run the test that requires 4 GPUs.

In the future, the internals of `PeriodicModelAveraging` can be simplified as an implementation of a specialized hierarchical model averaging, where `period_group_size_dict` only has a pair of period and world size.

### CODE REVIEW

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+                                Particularly, at each iteration, there will be at most a single
+                                process group that runs averaging -- the period of such group should
+                                have the largest period which the current step can be divided by.
+                                For example, if the dict has three keys: 2, 4, and 8,
+                                then this means totally three process groups will be created to
+                                average parameters every 2, 4, and 8 iterations, respectively.
+                                At the 4th iteration, only the second process group will run
+                                averaging, because the first process group should be a
+                                subset of the second process group, and no need to execute the first
+                                process proup redundantly.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:36]: s/proup/group

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                                On the other hand, the third process group can only be triggered
+                                every 8 iterations, so it will not be triggered at the 4th iteration.
+                                If ``None``, the default process group,
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:39]: so if the input for this is `None`, can we raise a warning saying that there's really no point ot using HierarchicalModelAverager with a `None` arg for this?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+                                which is created by :func:`torch.distributed.init_process_group`
+                                will be used, and it will run averaging every iteration, which is
+                                normally equivalent to the vanilla gradient averaging just like
+                                :class:`~torch.nn.DistributedDataParallel` (DDP). (default: ``None``)
+        warmup_steps (int): The number of warm-up steps. During this stage, model averaging is skipped.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:44]: should the warmup step period match with post_localSGD hook?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        >>>  # Assume that each machine has 4 GPUs, then each intra-machine subgroup has a size of 4.
+        >>>  subgroup, subgroups = dist.new_subgroups()
+        >>>  state = PostLocalSGDState(subgroup=subgroup, start_localSGD_iter=100)
+        >>>  model.register_comm_hook(state, post_localSGD_hook)
+        >>>
+        >>>  # Average parameters among each group of 8 processes every 4 iterations, and among all
+        >>>  # the 16 processes every 16 iterations.
+        >>>  averager = hierarchicalSGD.HierarchicalModelAverager(
+        >>>      period_group_size_dict=OrderedDict([(4, 8), (16, 16)]), warmup_steps=100)
+        >>>  # Note that ``warmup_steps`` must be the same as ``start_localSGD_iter`` used in ``PostLocalSGDState``.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:77]: do we check/assert against this at all? what can happen if it is not, do we have silent correctness/failure issues?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        >>>  # In the first 100 steps, run global gradient averaging like normal DDP at every step.
+        >>>  # After 100 steps, run model averaging at two levels.
+        >>>  for step in range(0, 200):
+        >>>     optimizer.zero_grad()
+        >>>     loss = loss_fn(output, labels)
+        >>>     loss.backward()
+        >>>     optimizer.step()
+        >>>     # Average parameters after ``optimizer.step()``.
+        >>>     # Thus, the inter-node communication only occurs periodically after ``warmup_steps``.
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:86]: specify what period it occurs at?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        >>>
+        >>>  # Average parameters among each group of 8 processes every 4 iterations, and among all
+        >>>  # the 16 processes every 16 iterations.
+        >>>  averager = hierarchicalSGD.HierarchicalModelAverager(
+        >>>      period_group_size_dict=OrderedDict([(4, 8), (16, 16)]), warmup_steps=100)
+        >>>  # Note that ``warmup_steps`` must be the same as ``start_localSGD_iter`` used in ``PostLocalSGDState``.
+        >>>  # In the first 100 steps, run global gradient averaging like normal DDP at every step.
+        >>>  # After 100 steps, run model averaging at two levels.
+        >>>  for step in range(0, 200):
+        >>>     optimizer.zero_grad()
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:81]: `optimizer` is not defined above

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        >>>     loss = loss_fn(output, labels)
+        >>>     loss.backward()
+        >>>     optimizer.step()
+        >>>     # Average parameters after ``optimizer.step()``.
+        >>>     # Thus, the inter-node communication only occurs periodically after ``warmup_steps``.
+        >>>     averager.average_parameters(model.parameters())
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:87]: does user have to call this for now or can it be integrated with an optimizer as you did for post local SGD?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 2 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:39]: Also, this doc doesn't appear consistent with the code which appears to raise an error if this arg is None?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    .. warning ::
+        `HierarchicalModelAverager` is experimental and subject to change.
+    """
+
+    def __init__(self, period_group_size_dict=None, warmup_steps=0, process_group=None):
+        if not period_group_size_dict:
+            raise ValueError("Arg ``period_group_size_dict`` must not be empty.")
+        self._periods = list(period_group_size_dict.keys())
+        if self._periods[0] < 0:
+            raise ValueError("The minimum period in arg ``period_group_size_dict`` must be a positive value.")
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:103]: `0` also doesn't seem to make sense, should the check be `if self.periods[0] <= 0`?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                "must be equal to the size of arg ``process_group``.")
+
+        self.period_process_group_dict = OrderedDict()
+        logger.info("Model averaging hierarchy:")
+        for period, group_size in period_group_size_dict.items():
+            logger.info(
+                f"\tEach group that has {group_size} processes average parameters every {period} iterations, "
+                "if no higher-level averaging.")
+            if group_size != overall_group_size:
+                self.period_process_group_dict[period], subgroups = dist.new_subgroups(
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:128]: `self.period_process_group_dict[period], _ = ... ` because `subgroups` is not used?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        and it can be divided by a period in the keys of ``period_process_group_dict``,
+        where ``step`` is increased by 1 at each iteration in the training loop.
+        If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
+        only the largest period is used, and the corresponding process group is used for averaging parameters.
+        """
+        if (
+            self.step >= self.warmup_steps
+            and self._is_process_group_found()
+        ):
+            utils.average_parameters(iter(params), self.group)
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:164]: unset `self.group` back to `None` after this so it is clear that it will need to be recomputed later?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        Returns whether ``step`` can be divided by a period in the keys of ``period_process_group_dict``.
+        If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
+        then sets set ``_group`` as the process group corresponding to the largest period,
+        since this process group will be used for averaging parameters at this ``step``.
+        """
+        for period in reversed(self._periods):
+            if self.step % period == 0:
+                self.group = self.period_process_group_dict[period]
+                return True
+        return False
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:150]: not the biggest fan of a function modifying state unless absolutely necessary, but fine with this as is.

One suggestion is that this function can return `Tuple[bool, Optional[ProcessGroup]]` where the latter is `None` if this function is returning `False`, otherwise it contains the process group to use. 

this will also help get rid of `self.group` entirely?

**Review by rohan-varma — COMMENTED** [review]
Thanks for contributing this Yi! Did an initial quick review, didn't get time to look at the tests yet and will circle back on those

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class HierarchicalModelAverager:
+    r"""
+    A group of model averagers used for hierarchical model averaging (hierarchical SGD).
+    Process groups of different sizes are organized in a hierarhicy, and they average parameters
+    by using different periods concurrently after the warm-up stage.
+    This is an extension of :class:`~torch.distributed.algorithms.model_averaging.averagers.PeriodicModelAverager`
```
[the remark this answers is not in the document]
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:16]: Yes, that's exactly what I added in the PR description. We can do some refactoring in a separate PR.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 3 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:44]: Yes, it's mentioned in L68.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        >>>  # Register a post-localSGD communication hook.
+        >>>  # Assume that each machine has 4 GPUs, then each intra-machine subgroup has a size of 4.
+        >>>  subgroup, subgroups = dist.new_subgroups()
+        >>>  state = PostLocalSGDState(subgroup=subgroup, start_localSGD_iter=100)
+        >>>  model.register_comm_hook(state, post_localSGD_hook)
+        >>>
+        >>>  # Average parameters among each group of 8 processes every 4 iterations, and among all
+        >>>  # the 16 processes every 16 iterations.
+        >>>  averager = hierarchicalSGD.HierarchicalModelAverager(
+        >>>      period_group_size_dict=OrderedDict([(4, 8), (16, 16)]), warmup_steps=100)
```
[the remark this answers is not in the document]
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:76]: This is the best API on my mind already, which can 1) hide the complexity of creating subgroups, 2) does not require any order between <period, group size> pairs. Since this API is experimental and subject to change, we can update the API once we have a better and concrete proposal.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 4 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:77]: Unfortunately not. If `warmup_steps` do not match, then the global gradient averaging and global model averaging can either have a gap or an overlap. The former can affect the convergence efficiency, and the latter will cause the unnecessary communication cost.

This issue is hard to prevent, unless there can be an API to query the state of DDP comm hook.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 6 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:81]: I am not sure if we really want to define this and `loss_fn` that clearly here.
The other similar comments in PyTorch do not define `optimizer` as well to be concise.
https://github.com/pytorch/pytorch/blob/dd1121435b14b48d862b5f16e4ce8b0d71f6dd5d/torch/optim/lr_scheduler.py#L1278
https://github.com/pytorch/pytorch/blob/12a1df27c73f05b8bfba72d68ac8f75f4e81dc35/torch/optim/swa_utils.py#L36

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 7 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:87]: We can integrate it with `PostLocalSGDOptimizer` later, which is more user-friendly. 

It should be noted that, one limitation of the optimizer API is that, if data loading or any step that involves communication and also overlaps with the optimizer step, then it can cause communication failure. Therefore, it's still necessary to provide the current way to use model averaging w/o involving an optimizer.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 5 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:86]: Already described in L73 when `HierarchicalModelAverager` is initialized.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 11 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:150]: Agree with your proposed change if we want to get rid of `self.group`. Actually I prefer keeping this state for now, as it can help me with some debugging. For example, even at a step that does not run model averaging, I still want to know what's the group size of the most recent model averaging.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 7 above)
Comment (rohan-varma) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:82]: Can you file an issue for doing this?

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+            raise ValueError("Arg ``period_group_size_dict`` must not be empty.")
+        self._periods = list(period_group_size_dict.keys())
+        if self._periods[0] <= 0:
+            raise ValueError("The minimum period in arg ``period_group_size_dict`` must be a positive value.")
+        elif self._periods[-1] == 1:
+            warnings.warn(
+                "When the maximum period in arg ``period_group_size_dict`` is 1, "
+                "no need to use model averaging because the communication cost "
+                "of all-reducing parameters will be no less than the cost of all-reducing gradients "
+                "by DistributedDataParall in the backward pass. Therefore, only "
```
Comment (rohan-varma) [inline torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:104]: nit: spelling

File: torch/testing/_internal/distributed/distributed_test.py [hunk 15]
```diff
@@ -1029,6 +1030,90 @@ def test_periodic_model_averager(self):
...
+                        # No model averaging, so the parameters are not updated.
+                        self.assertEqual(param.data, tensor)
+
+        @sandcastle_skip_if(
+            BACKEND not in DistTestCases.backend_feature["subgroup"],
+            f"The {BACKEND} backend does not support creating subgroups on CUDA devices"
+        )
+        @require_world_size(4)
+        @skip_if_lt_x_gpu(4)
+        def test_2_level_hierarchical_model_averager(self):
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:1075]: Can we add a test for 4 hierarchies as well or file an issue to do this?

**Review by rohan-varma — APPROVED** [review]
Thanks for addressing all the comments! Looks good, added ci/all to make sure all CI is run.

File: torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py (hunk 7 above)
Comment (wayi1) [reply torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py:82]: Filed https://github.com/pytorch/pytorch/issues/73382

File: torch/testing/_internal/distributed/distributed_test.py (hunk 15 above)
Comment (wayi1) [reply torch/testing/_internal/distributed/distributed_test.py:1075]: Do you mean creating another subgroup of size 3 besides the subgroup of size 2 and the global group?

**Comment by rohan-varma** [comment]
failures are unerlated
